gcovr/configuration.py

```diff
@@ -724,72 +730,106 @@ def parse_config_file(
-            raise error("semicolon comment ; ... is reserved")
-
-        if line.isspace() or not line:  # skip empty lines
-            continue
-
-        match = CONFIG_KV.match(line)
-        if not match:
-            raise error('expected "key = value" entry')
-
-        key: str = match.group(1).strip()
-        value: str = match.group(2)
-
-        if value.startswith('"'):
-            raise error('leading quote " is reserved')
-        if value.startswith("'"):
-            raise error("leading quote ' is reserved")
-        if value.endswith("\\"):
-            raise error("trailing backslash \\ is reserved")
-        if CONFIG_POSSIBLE_VARIABLE.search(value):
-            raise error(
-                "variable substitution syntax ({example}) is reserved",
-                example="${var}, $(var), or $var",
-            )
+    def get_configuration_iterator():
+        for lineno, line in enumerate(open_file, first_lineno):
+            line = line.rstrip()
+
+            def error(pattern: str, *args, **kwargs):
+                # pylint: disable=cell-var-from-loop
+                message = pattern.format(*args, **kwargs)
+                message += "\non this line: " + line
+                return SyntaxError(": ".join([filename, str(lineno), message]))
+
+            # strip (trailing) comments
+            line = CONFIG_HASH_COMMENT.sub("", line)
+
+            if CONFIG_SEMICOLON_COMMENT.search(line):
+                raise error("semicolon comment ; ... is reserved")
+
+            if line.isspace() or not line:  # skip empty lines
+                continue
+
+            match = CONFIG_KV.match(line)
+            if not match:
+                raise error('expected "key = value" entry')
+
+            key: str = match.group(1).strip()
+            value: str = match.group(2)
+
+            if value.startswith('"'):
+                raise error('leading quote " is reserved')
+            if value.startswith("'"):
+                raise error("leading quote ' is reserved")
+            if value.endswith("\\"):
+                raise error("trailing backslash \\ is reserved")
+            if CONFIG_POSSIBLE_VARIABLE.search(value):
+                raise error(
+                    "variable substitution syntax ({example}) is reserved",
+                    example="${var}, $(var), or $var",
+                )
 
-        yield ConfigEntry(key, value, filename=filename, lineno=lineno)
+            yield ConfigEntry(key, value, filename=filename, lineno=lineno)
 
+    return [parse_config_into_dict(get_configuration_iterator(), all_options)]
+
+def parse_toml_config_file(
+    open_file: BinaryIO,
+    filename: str,
+    all_options: Iterable[GcovrConfigOption] = None,
+) -> Iterable[ConfigEntry]:
+    r"""
+    Parse an toml-style configuration format.
+
+    Returns list of configurations.
+    """
+
+    configurations = []
+    toml_dict = tomllib.load(open_file)
+    if "gcovr" in toml_dict:
```

Comment:
I like the ConfigEntry because there the checking for valid keys is done and you get an error if an unknown option is used. I used your code here as a starting point.